src/policy/contract_policy.cpp

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2022 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+
+#include <policy/contract_policy.h>
+
+#include <coins.h>
+#include <consensus/amount.h>
+#include <logging.h>
+#include <tinyformat.h>
+
+#include <numeric>
+#include <vector>
+
+std::optional<std::pair<uint256, uint256>> CheckV3Inheritance(const Package& package)
+{
+    assert(std::all_of(package.cbegin(), package.cend(), [](const auto& tx){return tx != nullptr;}));
+    std::unordered_map<uint256, uint256, SaltedTxidHasher> v3_txid_to_wtxid;
+    for (const auto& tx : package) {
+        if (tx->nVersion == 3) {
+            v3_txid_to_wtxid.emplace(tx->GetHash(), tx->GetWitnessHash());
+        } else {
+            for (const auto& input : tx->vin) {
+                if (auto it = v3_txid_to_wtxid.find(input.prevout.hash); it != v3_txid_to_wtxid.end()) {
+                    return std::make_pair(it->second, tx->GetWitnessHash());
+                }
+            }
+        }
+    }
+    return std::nullopt;
+}
+
+std::optional<std::string> CheckV3Inheritance(const CTransactionRef& ptx,
+                                              const CTxMemPool::setEntries& ancestors)
+{
+    const auto version{ptx->nVersion};
+    if (version == 3) return std::nullopt;
+    for (const auto& entry : ancestors) {
+        if (entry->GetTx().nVersion == 3) {
+            return strprintf("tx that spends from %s must be nVersion=3",
+                             entry->GetTx().GetWitnessHash().ToString());
+        }
+    }
+    return std::nullopt;
+}
+
+CTxMemPool::setEntries GetV3Ancestors(const CTxMemPool::setEntries& ancestors)
+{
+    CTxMemPool::setEntries v3_ancestors;
+    for (const auto& entry : ancestors) {
+        if (entry->GetTx().nVersion == 3) {
+            v3_ancestors.insert(entry);
+        }
+    }
+    return v3_ancestors;
+}
+
+std::optional<std::string> ApplyV3Rules(const CTransactionRef& ptx,
+                                        const CTxMemPool::setEntries& ancestors,
+                                        const std::set<uint256>& direct_conflicts)
+{
+    // These rules only apply to transactions with nVersion=3.
+    if (ptx->nVersion != 3) return std::nullopt;
```

Comment:
Shouldn't we call `CheckV3Inheritance` here instead of letting the caller check that before? There are no cases where we want to call `ApplyV3Rules` without also checking v3 inheritance, are there?